// cql-execution/lib/datatypes/ratio.js
// Generated by CoffeeScript 1.12.7
(function() {
  var Ratio;

  module.exports.Ratio = Ratio = (function() {
    function Ratio(numerator, denominator) {
      this.numerator = numerator;
      this.denominator = denominator;
      if (this.numerator == null) {
        throw new Error("Cannot create a ratio with an undefined numerator");
      }
      if (this.denominator == null) {
        throw new Error("Cannot create a ratio with an undefined denominator");
      }
    }

    Object.defineProperties(Ratio.prototype, {
      isRatio: {
        get: function() {
          return true;
        }
      }
    });

    Ratio.prototype.clone = function() {
      return new Ratio(this.numerator.clone(), this.denominator.clone());
    };

    Ratio.prototype.toString = function() {
      return (this.numerator.toString()) + " : " + (this.denominator.toString());
    };

    Ratio.prototype.equals = function(other) {
      var divided_other, divided_this;
      if (other != null ? other.isRatio : void 0) {
        divided_this = this.numerator.dividedBy(this.denominator);
        divided_other = other.numerator.dividedBy(other.denominator);
        return divided_this.equals(divided_other);
      } else {
        return false;
      }
    };

    Ratio.prototype.equivalent = function(other) {
      var equal;
      equal = this.equals(other);
      if (equal == null) {
        return false;
      }
      return equal;
    };

    return Ratio;

  })();

}).call(this);
